### DISCUSSION TITLE
(Regional Edge Functions) Allow Multiple regions to be selected

### DISCUSSION THREAD

**Discussion by IamManchanda** [vercel/community#974] [Idea]
This is a great addition for Next.js & Vercel https://vercel.com/blog/regional-execution-for-ultra-low-latency-rendering-at-the-edge but let's suppose I have a Data Fetching Need where

- I will be using globally distributed database like [Amazon Aurora Global Database](https://aws.amazon.com/rds/aurora/global-database/), or [Google Cloud Spanner](https://cloud.google.com/spanner) etc 
- I will be selecting Multiple Regions, suppose: USA, Ireland, India

In that case, can't this be done ⬇️ ?

```js
export const config = {
  runtime: 'edge', // 'experimental-edge' as of today
  regions: ['iad1', 'dub1', 'bom1'] // https://vercel.com/docs/concepts/edge-network/regions#region-list
};
```

![vercel-multi-region](https://user-images.githubusercontent.com/4970624/198821895-15df5cb3-098c-49bc-8e09-8433fe79a0bd.png)

**Comment by amyegan** [comment]
Hi @IamManchanda! All Vercel accounts can [change the default region of their Edge Functions](https://vercel.com/docs/concepts/functions/serverless-functions/regions) to a specific region to be near their data source. But only Enterprise accounts have the ability to [use multiple Serverless Function regions](https://vercel.com/docs/concepts/functions/serverless-functions/regions#multiple-regions). 

If you need this feature and want to upgrade to an Enterprise plan, please contact our [Enterprise department](https://vercel.com/contact/sales) for help tailoring a plan to meet your needs.

**Reply by IamManchanda** [reply]
No I meant with Developer version (Free or Pro version) not Enterprise... 

To my best of knowledge, Edge Functions support multiple regions, ie Edge Computing for all accounts, free and pro not just enterprise... Wasn't expecting "contact enterprise department" response but anyways thanks for your response @amyegan 🙌

**Reply by amyegan** [reply]
Right, Edge Functions can be either [global or regional](https://vercel.com/changelog/regional-edge-functions-are-now-available). They run globally in every Vercel region by default. Hobby and Pro plans have the option to select one region to compute closer to their database. Only Enterprise plans have the ability to select multiple regions.

>Deploying to multiple regions (or all) is limited to Enterprise plans, but Pro and Hobby plans can select any single region.
>— [regions config documentation](https://vercel.com/docs/project-configuration#project-configuration/regions)

If I am misunderstanding what you are asking, please let me know and I will work with the larger team to get you the answer. 🙏

**Reply by IamManchanda** [reply]
@amyegan cc @rauchg @leerob - Hello first of all, Great work with Next.js 13 & my best wishes for Next.js 14 🙌 

What I am trying to say is that **please provide the developers** (with **Pro at least** if not possible with Free Hobby) the **ability to choose multiple regions (besides all regions or single region)** for **"Edge Functions" for the [Edge Runtime](https://beta.nextjs.org/docs/rendering/edge-and-nodejs-runtimes)** (not talking about Serverless functions for the Node Runtime)

> Right, Edge Functions can be either [global or regional](https://vercel.com/changelog/regional-edge-functions-are-now-available)

What I am looking for is "multi-regional" (for Pro & Above, if not Hobby & Above) besides "global" or "regional".

---

The thing is that these days to **go global cloud compute**, the Backend DB (or Backend API) i.e the Data is usually distributed with Multi Region Cloud Service where you distribute your data to multiple regions 
with Cloud services like ⬇️  (External Cloud Service outside Vercel Platform for data needs)
 
 - Amazon [Aurora Global Database](https://aws.amazon.com/rds/aurora/global-database/) or [DynamoDB Global Tables](https://aws.amazon.com/dynamodb/global-tables/) or [Global Accelerator](https://aws.amazon.com/global-accelerator/)
 - Google [Cloud Spanner](https://cloud.google.com/spanner) or [Network Service Tiers](https://cloud.google.com/network-tiers)
 - [Fly.io](https://fly.io/) `flyctl regions add <list of regions>`
 - MongoDB Atlas's [Multi-Cloud, Multi-Region & Workload Isolation](https://www.mongodb.com/docs/atlas/cluster-config/multi-cloud-distribution/)
  
**Now let's suppose** I am using any one of these above services for my data requirements (Backend DB and/or Backend API) and using Cloud Provider like AWS or Google, I have selected three Cloud regions for my data:
 - Washington DC, USA (`iad1`)
 - Dublin, Ireland (`dub1`)
 - Mumbai, India (`bom1`)

I would want the **Next.js's Edge API Route Backend to live in those regions only** instead of All Regions, let me tell you why...

Let's suppose the **User is coming at my Web App from Sydney, Australia (`syd1`) or nearby** 

Then, in that case I would want my request response cycle from "Vercel to Cloud Backend/Database Provider" to be `bom1` to `bom1` and not `syd1` to `bom1` as explained in this below image.

![vercel-multi-region](https://user-images.githubusercontent.com/4970624/198817096-476c202c-0a82-439c-b54a-32bdccf6668a.png)


And thus better to have something like this ⬇️ 

```js
import { type NextRequest } from 'next/server'

export const config = {
  runtime: 'edge', // 'experimental-edge' as of today
  regions: ['iad1', 'dub1', 'bom1'] // https://vercel.com/docs/concepts/edge-network/regions#region-list
}

export default async function handler(req: NextRequest) {
  // Below example is for External Service Backend API access but 
  // similar pattern can be followed for External Backend DB access also, 
  // let's say with Prisma https://www.prisma.io/nextjs
  const data = await fetch(`https://api-multi-region-external-service.com/data`, {
    method: req.method,
    headers: {
       authorization: process.env.API_KEY,
    },
  });
  
  // ... other code
}
```
 
or even better (if possible) like this ⬇️ where we have an ability to grab the **current edge region**  from `NextRequest`

```js
import { type NextRequest } from 'next/server'

export const config = {
  runtime: 'edge', // 'experimental-edge' as of today
  regions: ['iad1', 'dub1', 'bom1'] // https://vercel.com/docs/concepts/edge-network/regions#region-list
}

export default async function handler(req: NextRequest) {
  // Below example is for External Service Backend API access but 
  // similar pattern can be followed for External Backend DB access also, 
  // let's say with Prisma https://www.prisma.io/nextjs
  const data = await fetch(`https://api-${req.region}-external-service.com/data`, {
    method: req.method,
    headers: {
      authorization: process.env.API_KEY,
    },
  });
  
  // ... other code
}
```

**Comment by IamManchanda** [comment]
@amyegan @rauchg @leerob - Oh so this got shipped, right?

<img width="672" alt="Screenshot 2023-05-04 at 11 47 58 PM" src="https://user-images.githubusercontent.com/4970624/236293954-7e824349-a332-4a85-a2be-00b5b3a28901.png">

**Reply by amyegan** [reply]
You've likely already tried it by now and found out for yourself. But I believe that, for the latest version of Next.js at least, this does mean that you should be able to set multiple regions 🙌
